File: dsa-problems/AdvancedGraphs/solutions/python/AGR-015-directed-cycle-basis.py
import sys
from collections import deque

def cycle_basis(n: int, edges: list[tuple[int, int]]) -> list[list[int]]:
    m = len(edges)
    adj = [[] for _ in range(n)]
    for i, (u, v) in enumerate(edges):
        adj[u].append((v, i))
        
    # Calc basis size
    undir_adj = [[] for _ in range(n)]
    for u, v in edges:
        undir_adj[u].append(v)
        undir_adj[v].append(u)
        
    visited = [False] * n
    c = 0
    for i in range(n):
        if not visited[i]:
            c += 1
            q = [i]
            visited[i] = True
            while q:
                u = q.pop()
                for v in undir_adj[u]:
                    if not visited[v]:
                        visited[v] = True
                        q.append(v)
    
    D = m - n + c
    
    basis = [None] * m
    result = []
    
    def get_path(start, target):
        if start == target: return []
        parent = [-1] * n
        parent_edge = [-1] * n
        q = deque([start])
        parent[start] = start
        
        while q:
            u = q.popleft()
            if u == target: break
            for v, idx in adj[u]:
                if parent[v] == -1:
                    parent[v] = u
                    parent_edge[v] = idx
                    q.append(v)
        
        if parent[target] == -1: return None
        
        path = []
        curr = target
        while curr != start:
            path.append(parent_edge[curr])
            curr = parent[curr]
        return path[::-1]

    for i in range(m):
        if len(result) == D: break
        
        u, v = edges[i]
        path = get_path(v, u)
        if path is None: continue
        
        # Vector
        vec = 0
        vec |= (1 << i)
        for idx in path:
            vec |= (1 << idx)
            
        # Insert
        temp_vec = vec
        inserted = False
        # Find pivot
        # We iterate bits from low to high or high to low?
        # Use low to high (0 to m-1)
        for bit in range(m):
            if (temp_vec >> bit) & 1:
                if basis[bit] is None:
                    basis[bit] = temp_vec
                    inserted = True
                    break
                else:
                    temp_vec ^= basis[bit]
        
        if inserted:
            cycle = [u, v]
            curr = v
            for idx in path:
                # Edge idx is x->y. curr is x.
                # We need y.
                next_node = edges[idx][1]
                cycle.append(next_node)
                curr = next_node
            result.append(cycle)
            
    return result

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        m = int(next(iterator))
        edges = []
        for _ in range(m):
            u = int(next(iterator))
            v = int(next(iterator))
            edges.append((u, v))
            
        cycles = cycle_basis(n, edges)
        out = [str(len(cycles))]
        for cyc in cycles:
            out.append(str(len(cyc)) + " " + " ".join(map(str, cyc)))
        sys.stdout.write("\n".join(out).strip())
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/AdvancedGraphs/solutions/java/AGR-015-directed-cycle-basis.java
import java.util.*;

class Solution {
    public List<List<Integer>> cycleBasis(int n, int[][] edges) {
        int m = edges.length;
        List<List<int[]>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int i = 0; i < m; i++) {
            adj.get(edges[i][0]).add(new int[]{edges[i][1], i});
        }

        // Calc basis size
        int c = 0;
        boolean[] visited = new boolean[n];
        List<List<Integer>> undirAdj = new ArrayList<>();
        for (int i = 0; i < n; i++) undirAdj.add(new ArrayList<>());
        for (int[] e : edges) {
            undirAdj.get(e[0]).add(e[1]);
            undirAdj.get(e[1]).add(e[0]);
        }
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                c++;
                dfs(i, visited, undirAdj);
            }
        }
        int D = m - n + c;

        BitSet[] basis = new BitSet[m]; // basis[i] has pivot at i
        List<List<Integer>> result = new ArrayList<>();
        
        for (int i = 0; i < m; i++) {
            if (result.size() == D) break;
            
            int u = edges[i][0];
            int v = edges[i][1];
            
            List<Integer> path = bfs(v, u, n, adj);
            if (path == null) continue;
            
            BitSet vec = new BitSet(m);
            vec.set(i);
            for (int eIdx : path) vec.set(eIdx);
            
            if (insert(basis, vec)) {
                List<Integer> cycle = new ArrayList<>();
                cycle.add(u);
                int curr = v;
                cycle.add(curr);
                for (int eIdx : path) {
                    curr = edges[eIdx][1];
                    cycle.add(curr);
                }
                result.add(cycle);
            }
        }
        return result;
    }
    
    private void dfs(int u, boolean[] visited, List<List<Integer>> adj) {
        visited[u] = true;
        for (int v : adj.get(u)) {
            if (!visited[v]) dfs(v, visited, adj);
        }
    }
    
    private List<Integer> bfs(int start, int target, int n, List<List<int[]>> adj) {
        if (start == target) return new ArrayList<>();
        int[] parentEdge = new int[n];
        int[] parentNode = new int[n];
        Arrays.fill(parentEdge, -1);
        Arrays.fill(parentNode, -1);
        
        Queue<Integer> q = new ArrayDeque<>();
        q.add(start);
        parentNode[start] = start;
        
        while (!q.isEmpty()) {
            int u = q.poll();
            if (u == target) break;
            for (int[] edge : adj.get(u)) {
                int v = edge[0];
                int idx = edge[1];
                if (parentNode[v] == -1) {
                    parentNode[v] = u;
                    parentEdge[v] = idx;
                    q.add(v);
                }
            }
        }
        
        if (parentNode[target] == -1) return null;
        
        List<Integer> path = new ArrayList<>();
        int curr = target;
        while (curr != start) {
            path.add(parentEdge[curr]);
            curr = parentNode[curr];
        }
        Collections.reverse(path);
        return path;
    }
    
    private boolean insert(BitSet[] basis, BitSet vec) {
        for (int i = vec.nextSetBit(0); i >= 0; i = vec.nextSetBit(i + 1)) {
            if (basis[i] == null) {
                basis[i] = (BitSet) vec.clone();
                return true;
            }
            vec.xor(basis[i]);
        }
        return false;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[][] edges = new int[m][2];
        for (int i = 0; i < m; i++) {
            edges[i][0] = sc.nextInt();
            edges[i][1] = sc.nextInt();
        }

        Solution solution = new Solution();
        List<List<Integer>> cycles = solution.cycleBasis(n, edges);
        StringBuilder sb = new StringBuilder();
        sb.append(cycles.size()).append('\n');
        for (List<Integer> cyc : cycles) {
            sb.append(cyc.size());
            for (int v : cyc) sb.append(' ').append(v);
            sb.append('\n');
        }
        System.out.print(sb.toString().trim());
        sc.close();
    }
}
